<a href="https://colab.research.google.com/github/sumaiyaa0821/ETE456/blob/main/1608021_Exp02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Experiment Name: Classification Project on Stroke Prediction Dataset**

**Course No:** ETE 456

**Name:** Sumaiya Binta Rowshon

**ID:** 1608021

**Objective:**

1.Apply various classification algorithms on a real world dataset.

In [1]:
import warnings
warnings.filterwarnings("ignore")

**Import Libraries **

In [2]:
import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt

In [3]:
# Download the data
!wget -O stroke-data.csv https://www.dropbox.com/s/zgburk3yces5tee/healthcare-dataset-stroke-data.csv?dl=0

--2021-12-20 14:03:10--  https://www.dropbox.com/s/zgburk3yces5tee/healthcare-dataset-stroke-data.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.81.18, 2620:100:6035:18::a27d:5512
Connecting to www.dropbox.com (www.dropbox.com)|162.125.81.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/zgburk3yces5tee/healthcare-dataset-stroke-data.csv [following]
--2021-12-20 14:03:11--  https://www.dropbox.com/s/raw/zgburk3yces5tee/healthcare-dataset-stroke-data.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uce2c6665473a862d4965d0f232f.dl.dropboxusercontent.com/cd/0/inline/BcPy38qxBp7r65DxXh739NfpiTskOCsUsv6K7y6npPnyTZj7Zoe8esQsa4TgoBV8tBFlIWMDaBtAXrmwL-cbuK3bqbLXYg5S0ut1PZcKthkIRYYOIgqz_Uye816aadh6Hzj-RhEw-t1RPoaDsvH48qTG/file# [following]
--2021-12-20 14:03:11--  https://uce2c6665473a862d4965d0f232f.dl.dropboxusercontent.com/cd/0/inline/BcPy38qxBp7r65Dx

In [4]:
"""importing the dataset """

dataset = pd.read_csv('stroke-data.csv')
dataset

id  gender   age  ...   bmi   smoking_status stroke
0      9046    Male  67.0  ...  36.6  formerly smoked      1
1     51676  Female  61.0  ...   NaN     never smoked      1
2     31112    Male  80.0  ...  32.5     never smoked      1
3     60182  Female  49.0  ...  34.4           smokes      1
4      1665  Female  79.0  ...  24.0     never smoked      1
...     ...     ...   ...  ...   ...              ...    ...
5105  18234  Female  80.0  ...   NaN     never smoked      0
5106  44873  Female  81.0  ...  40.0     never smoked      0
5107  19723  Female  35.0  ...  30.6     never smoked      0
5108  37544    Male  51.0  ...  25.6  formerly smoked      0
5109  44679  Female  44.0  ...  26.2          Unknown      0

[5110 rows x 12 columns]

In [5]:
features = dataset[['age',	'hypertension',	'heart_disease',	'bmi','Residence_type','avg_glucose_level',	'smoking_status']]  # for independent
target = dataset[['stroke']]   # for dependent

In [6]:
features

age  hypertension  ...  avg_glucose_level   smoking_status
0     67.0             0  ...             228.69  formerly smoked
1     61.0             0  ...             202.21     never smoked
2     80.0             0  ...             105.92     never smoked
3     49.0             0  ...             171.23           smokes
4     79.0             1  ...             174.12     never smoked
...    ...           ...  ...                ...              ...
5105  80.0             1  ...              83.75     never smoked
5106  81.0             0  ...             125.20     never smoked
5107  35.0             0  ...              82.99     never smoked
5108  51.0             0  ...             166.29  formerly smoked
5109  44.0             0  ...              85.28          Unknown

[5110 rows x 7 columns]

In [7]:
target

stroke
0          1
1          1
2          1
3          1
4          1
...      ...
5105       0
5106       0
5107       0
5108       0
5109       0

[5110 rows x 1 columns]

In [8]:
from sklearn.impute import SimpleImputer

In [9]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(features[['age',	'hypertension',	'heart_disease','bmi','avg_glucose_level']])

In [11]:
features[['age',	'hypertension',	'heart_disease','bmi','avg_glucose_level']]= imputer.transform(features[['age',	'hypertension',	'heart_disease','bmi','avg_glucose_level']])

In [12]:
features

age  hypertension  ...  avg_glucose_level   smoking_status
0     67.0           0.0  ...             228.69  formerly smoked
1     61.0           0.0  ...             202.21     never smoked
2     80.0           0.0  ...             105.92     never smoked
3     49.0           0.0  ...             171.23           smokes
4     79.0           1.0  ...             174.12     never smoked
...    ...           ...  ...                ...              ...
5105  80.0           1.0  ...              83.75     never smoked
5106  81.0           0.0  ...             125.20     never smoked
5107  35.0           0.0  ...              82.99     never smoked
5108  51.0           0.0  ...             166.29  formerly smoked
5109  44.0           0.0  ...              85.28          Unknown

[5110 rows x 7 columns]

In [13]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(target[['stroke']])

In [14]:
target[['stroke']]= imputer.transform(target[['stroke']])

In [15]:
target

stroke
0        1.0
1        1.0
2        1.0
3        1.0
4        1.0
...      ...
5105     0.0
5106     0.0
5107     0.0
5108     0.0
5109     0.0

[5110 rows x 1 columns]

In [16]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

In [17]:
labelencoder_f = LabelEncoder()
#the country column is represented by numeric value
features['Residence_type'] = labelencoder_f.fit_transform(features['Residence_type'])

In [18]:
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(features[['smoking_status']]))

In [19]:
encoded_labels.columns = encoder.get_feature_names(['smoking_status'])
dataset= pd.concat([features, encoded_labels ], axis=1)

In [20]:
dataset

age  hypertension  ...  smoking_status_never smoked  smoking_status_smokes
0     67.0           0.0  ...                          0.0                    0.0
1     61.0           0.0  ...                          1.0                    0.0
2     80.0           0.0  ...                          1.0                    0.0
3     49.0           0.0  ...                          0.0                    1.0
4     79.0           1.0  ...                          1.0                    0.0
...    ...           ...  ...                          ...                    ...
5105  80.0           1.0  ...                          1.0                    0.0
5106  81.0           0.0  ...                          1.0                    0.0
5107  35.0           0.0  ...                          1.0                    0.0
5108  51.0           0.0  ...                          0.0                    0.0
5109  44.0           0.0  ...                          0.0                    0.0

[5110 rows x 11 columns]

In [21]:
dataset.columns

Index(['age', 'hypertension', 'heart_disease', 'bmi', 'Residence_type',
       'avg_glucose_level', 'smoking_status', 'smoking_status_Unknown',
       'smoking_status_formerly smoked', 'smoking_status_never smoked',
       'smoking_status_smokes'],
      dtype='object')

In [22]:
new_features = dataset[['age', 'hypertension', 'heart_disease', 'bmi',
       'avg_glucose_level','Residence_type',
       'smoking_status_Unknown',
       'smoking_status_formerly smoked', 'smoking_status_never smoked',
       'smoking_status_smokes']]

In [23]:
from sklearn.model_selection import train_test_split

In [24]:
"""Spliting the Dataset into Training Set and Test Set """

X_train,X_test,y_train,y_test=train_test_split(new_features,target,test_size=0.3,random_state=0)

In [25]:
print(X_train.shape)
print(X_test.shape)

(3577, 10)
(1533, 10)


In [26]:
X_test

age  hypertension  ...  smoking_status_never smoked  smoking_status_smokes
42    82.0           0.0  ...                          0.0                    1.0
380    4.0           0.0  ...                          0.0                    0.0
3524  58.0           0.0  ...                          1.0                    0.0
352   20.0           0.0  ...                          1.0                    0.0
4927  10.0           0.0  ...                          0.0                    0.0
...    ...           ...  ...                          ...                    ...
929   75.0           0.0  ...                          0.0                    0.0
2629  45.0           0.0  ...                          0.0                    0.0
663   38.0           0.0  ...                          1.0                    0.0
2303  75.0           0.0  ...                          0.0                    1.0
3122  57.0           1.0  ...                          0.0                    1.0

[1533 rows x 10 columns]

In [27]:
from sklearn.linear_model import LogisticRegression

# Fitting Logistic Regression to the training dataset
lr = LogisticRegression()

lr.fit(X_train,y_train)

LogisticRegression()

In [28]:
# prediction
y_pred = lr.predict(X_test)

In [29]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score,f1_score,precision_score,recall_score

In [30]:
# Making confusing matrix (actual,prediction)
confusion_matrix(y_test,y_pred)

array([[1457,    0],
       [  76,    0]])

In [31]:
# Making confusing matrix
# it is used to check the accuracy of the classification
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      1.00      0.97      1457
         Yes       0.00      0.00      0.00        76

    accuracy                           0.95      1533
   macro avg       0.48      0.50      0.49      1533
weighted avg       0.90      0.95      0.93      1533



In [32]:
from sklearn.neighbors import KNeighborsClassifier

# Classifier Model
classifier = KNeighborsClassifier(n_neighbors=3, metric = 'minkowski')
classifier.fit(X_train,y_train)
# Prediction
y_pred = classifier.predict(X_test)

In [33]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

[[1431   26]
 [  71    5]]


In [34]:
# Making confusing matrix
# it is used to check the accuracy of the classification
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      0.98      0.97      1457
         Yes       0.16      0.07      0.09        76

    accuracy                           0.94      1533
   macro avg       0.56      0.52      0.53      1533
weighted avg       0.91      0.94      0.92      1533



In [35]:
from sklearn.svm import SVC

# Classifier Model
classifier = SVC(kernel = 'linear', random_state = 42)
classifier.fit(X_train,y_train)
# Prediction
y_pred = classifier.predict(X_test)

In [36]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

[[1457    0]
 [  76    0]]


In [37]:
# Making confusing matrix
# it is used to check the accuracy of the classification
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      1.00      0.97      1457
         Yes       0.00      0.00      0.00        76

    accuracy                           0.95      1533
   macro avg       0.48      0.50      0.49      1533
weighted avg       0.90      0.95      0.93      1533



In [38]:
from sklearn.tree import DecisionTreeClassifier
# Classifier Model
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train,y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

In [39]:
# Prediction
y_pred = classifier.predict(X_test)

In [40]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

[[1395   62]
 [  64   12]]


In [41]:
# Making confusing matrix
# it is used to check the accuracy of the classification
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.96      0.96      0.96      1457
         Yes       0.16      0.16      0.16        76

    accuracy                           0.92      1533
   macro avg       0.56      0.56      0.56      1533
weighted avg       0.92      0.92      0.92      1533



In [42]:
from sklearn.ensemble import RandomForestClassifier
# Classifier Model
classifier = RandomForestClassifier(n_estimators=32, criterion ='entropy', random_state = 40)
classifier.fit(X_train,y_train)

RandomForestClassifier(criterion='entropy', n_estimators=32, random_state=40)

In [43]:
# Prediction
y_pred = classifier.predict(X_test)

In [44]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

[[1456    1]
 [  74    2]]


In [45]:
# Making confusing matrix
# it is used to check the accuracy of the classification
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.95      1.00      0.97      1457
         Yes       0.67      0.03      0.05        76

    accuracy                           0.95      1533
   macro avg       0.81      0.51      0.51      1533
weighted avg       0.94      0.95      0.93      1533



**Result Analysis: **

In this project, different classification algorithm was used to classify or predict stroke from dataset. The evaluation matrices to evaluate classification algorithms are Precision, Recall, F1 Score and Accuracy. Here for comparision purposes, the weighted average of models with accuracy will be considered.

 **Logistic Regression **

Weighted Average Precision : 90%
Weighted Average Recall : 95%
Weighted Average F1-Score : 93%*italicised text*
Accuracy : 95%


**KNN Classifier:**

Weighted Average Precision : 91%
Weighted Average Recall : 94%
Weighted Average F1-Score : 92%
Accuracy : 94%

**Support Vector Machine:**
Weighted Average Precision : 90%
Weighted Average Recall : 95%
Weighted Average F1-Score : 93%
Accuracy : 95%

**Decision Tree Classification**

Weighted Average Precision : 92%
Weighted Average Recall : 92%
Weighted Average F1-Score : 92%
Accuracy : 92%

**Random Forest Classification**

Weighted Average Precision : 94%
Weighted Average Recall : 95%
Weighted Average F1-Score : 93%
Accuracy : 95%

**Discussion:**

Several classifiers are employed in this research to predict Stroke from the "Stroke Dataset." Logistic Regression, K-Nearest Neighbor, Support Vector Machine, Decision Tree, and Random Forest Classifier are the classifiers utilized in this research. The procedure libraries were first imported, and then the dataset was downloaded from Dropbox. Then there was some data pre-processing, such as encoding (label encoding, one hot encoding), feature scaling, and dealing with missing files. After that, the dataset was partitioned and the classifiers were trained. The classifiers were then evaluated using some evaluation matrices, and the comparison based on evaluation was displayed in the Result Analysis. The Random Forest Classifier has the highest evaalution values, according to the comparison.
